src/azul/service/responseobjects/hca_response_v5.py

```diff
@@ -281,47 +292,139 @@ def _construct_tsv_content(self):
+        header_row[first_col_idx] = 'entity:bundle_id'  # first column headers needs to have that format
+        header_row = self._set_as_first_col(first_col_idx, header_row)
+        col_idx = header_row.index('bundles_version')
+        header_row[col_idx] = 'bundle_version'
+        tsv_writer.writerow(header_row)
+        all_bundles = defaultdict()
+        for hit_obj in es_search.scan():
+            hit = hit_obj.to_dict()
+            data_template = deepcopy(bundles)
+            bundle_key = first(hit['bundles'])['uuid']
+            if bundle_key in all_bundles.keys():
+                all_bundles[bundle_key] = self._put_bdbag_tsv_file_data_in_dict(hit, all_bundles[bundle_key])
+            else:
+                all_bundles[bundle_key] = self._put_bdbag_tsv_file_data_in_dict(hit, data_template)
+        for data in all_bundles.values():
+            data_row = self._construct_bdbag_tsv_file_data_row(data)
+            data_row = self._set_as_first_col(first_col_idx, data_row)
+            tsv_writer.writerow(data_row)
 
-        participant_file_object = StringIO()
-        participant_writer = csv.writer(participant_file_object, dialect='excel-tab')
-        participant_writer.writerow(['entity:participant_id'])
-        participant_writer.writerows(zip(participants))
-        return self._create_zipped_bdbag(participant_file_object, sample_file_object)
+        return self._create_zipped_bdbag(tsv_file_object)
 
     @staticmethod
-    def _create_zipped_bdbag(participant_file_object: StringIO, sample_file_object: StringIO) -> str:
+    def _set_as_first_col(first_col_idx: int, row: list) -> list:
+        row[0], row[first_col_idx] = row[first_col_idx], first(row)
+
+        return row
+
+    @staticmethod
+    def _create_bdbag_tsv_file_header_row(data: BDBagOrderedDict) -> list:
+        row = []
+        for entity, value_dict in data.items():
+            for value_key in value_dict.keys():
+                row.append(entity + '_' + value_key)
+
+        return row
+
+    @staticmethod
+    def _collect_bdbag_tsv_file_headers(hit: dict, data: BDBagOrderedDict) -> BDBagOrderedDict:
+        for contents in hit.keys():
+            # Bundle ID and version.
+            if isinstance(hit[contents], list):
+                if len(hit[contents]) > 1:
+                    bundles_containing_file = {'bundle_uuid': bundle['uuid'] for bundle in hit[contents]}
+                    logger.info(f'File found in more than one bundle ID: '
+                                   f'{bundles_containing_file} - processing first encountered.')
+                data[contents] = {key: '' for key in first(hit[contents]).keys()}
+            # File and metadata attributes.
+            elif isinstance(hit[contents], dict):
+                for entity in hit[contents].keys():
+                    if entity == 'files':
+                        file_type = first(first(hit[contents][entity])['file_format'].split('.'))
```

Comment:
No. You can't just discard everything after the first period in the format.

Some projects use `fastq`, while others use `fastq.gz` but discarding the part after a period is not how we would handle this as it would also affect other types that we know nothing about. We just need to `if file_format == "fastq.gz": file_format = "fastq"`.

And you're falling back in to the defensive style of coding again. Like using `first()` instead of `one()`.